diff --git a/freemarker/core/FMParser.java b/freemarker/core/FMParser.java
index 81cb29b..e19bbfd 100644
--- a/freemarker/core/FMParser.java
+++ b/freemarker/core/FMParser.java
@@ -65,7 +65,7 @@ public class FMParser implements FMParserConstants {
      */
     private int continuableDirectiveNesting;
 
-    private boolean inMacro, inFunction;
+    private boolean inMacro, inFunction, requireArgsSpecialVariable;
     private LinkedList escapes = new LinkedList();
     private int mixedContentNesting; // for stripText
 
@@ -392,7 +392,7 @@ public class FMParser implements FMParserConstants {
      */
     private void notBooleanLiteral(Expression exp, String expected) throws ParseException {
         if (exp instanceof BooleanLiteral) {
-            throw new ParseException("Found: " + exp.getCanonicalForm() + ". Expecting " + expected, exp);
+            throw new ParseException("Found: " + exp.getCanonicalForm() + " literal. Expecting " + expected , exp);
         }
     }
 
@@ -450,7 +450,7 @@ public class FMParser implements FMParserConstants {
         notNumberLiteral(exp, "boolean (true/false)");
     }
 
-    private Expression escapedExpression(Expression exp) {
+    private Expression escapedExpression(Expression exp) throws ParseException {
         if (!escapes.isEmpty()) {
             return ((EscapeBlock) escapes.getFirst()).doEscape(exp);
         } else {
@@ -884,10 +884,10 @@ result = lhs;
         throw new ParseException();
       }
       rhs = RelationalExpression();
-notHashLiteral(lhs, "string");
-                notHashLiteral(rhs, "string");
-                notListLiteral(lhs, "string");
-                notListLiteral(rhs, "string");
+notHashLiteral(lhs, "different type for equality check");
+                notHashLiteral(rhs, "different type for equality check");
+                notListLiteral(lhs, "different type for equality check");
+                notListLiteral(rhs, "different type for equality check");
                 result = new ComparisonExpression(lhs, rhs, t.image);
                 result.setLocation(template, lhs, rhs);
     } else {
@@ -933,12 +933,8 @@ result = lhs;
         throw new ParseException();
       }
       rhs = RangeExpression();
-notHashLiteral(lhs, "number");
-            notHashLiteral(rhs, "number");
-            notListLiteral(lhs, "number");
-            notListLiteral(rhs, "number");
-            notStringLiteral(lhs, "number");
-            notStringLiteral(rhs, "number");
+numberLiteralOnly(lhs);
+            numberLiteralOnly(rhs);
             result = new ComparisonExpression(lhs, rhs, t.image);
             result.setLocation(template, lhs, rhs);
     } else {
@@ -1135,6 +1131,13 @@ BuiltinVariable result = null;
             parseTimeValue = new SimpleScalar(outputFormat.getName());
         } else if (nameStr.equals(BuiltinVariable.AUTO_ESC) || nameStr.equals(BuiltinVariable.AUTO_ESC_CC)) {
             parseTimeValue = autoEscaping ? TemplateBooleanModel.TRUE : TemplateBooleanModel.FALSE;
+        } else if (nameStr.equals(BuiltinVariable.ARGS)) {
+            if (!inMacro && !inFunction) {
+                {if (true) throw new ParseException("The \"" + BuiltinVariable.ARGS + "\" special variable must be "
+                        + "inside a macro or function in the template source code.", template, name);}
+            }
+            requireArgsSpecialVariable = true;
+            parseTimeValue = null;
         } else {
             parseTimeValue = null;
         }
@@ -1576,8 +1579,8 @@ result.setLocation(template, t, t);
 
   final public HashLiteral HashLiteral() throws ParseException {Token begin, end;
     Expression key, value;
-    ArrayList keys = new ArrayList();
-    ArrayList values = new ArrayList();
+    ArrayList<Expression> keys = new ArrayList<Expression>();
+    ArrayList<Expression> values = new ArrayList<Expression>();
     begin = jj_consume_token(OPENING_CURLY_BRACKET);
     switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
     case STRING_LITERAL:
@@ -1652,7 +1655,9 @@ stringLiteralOnly(key);
       ;
     }
     end = jj_consume_token(CLOSING_CURLY_BRACKET);
-HashLiteral result = new HashLiteral(keys, values);
+keys.trimToSize();
+        values.trimToSize();
+        HashLiteral result = new HashLiteral(keys, values);
         result.setLocation(template, begin, end);
         {if ("" != null) return result;}
     throw new Error("Missing return statement in function");
@@ -2697,17 +2702,16 @@ LibraryLoad result = new LibraryLoad(template, nameExp, ns.image);
   final public Macro Macro() throws ParseException {Token arg, start, end;
     Expression nameExp;
     String name;
-    ArrayList argNames = new ArrayList();
-    HashMap args = new HashMap();
-    ArrayList defNames = new ArrayList();
+    Map<String, Expression> paramNamesWithDefault = new LinkedHashMap<String, Expression>();
     Expression defValue = null;
+    String catchAllParamName = null;
+    boolean isCatchAll = false;
     List lastIteratorBlockContexts;
     int lastBreakableDirectiveNesting;
-    int lastContiunableDirectiveNesting;
+    int lastContinuableDirectiveNesting;
     TemplateElements children;
-    boolean isFunction = false, hasDefaults = false;
-    boolean isCatchAll = false;
-    String catchAll = null;
+    boolean isFunction = false;
+    boolean hasDefaults = false;
     switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
     case MACRO:{
       start = jj_consume_token(MACRO);
@@ -2727,6 +2731,7 @@ if (inMacro || inFunction) {
             {if (true) throw new ParseException("Macro or function definitions can't be nested into each other.", template, start);}
         }
         if (isFunction) inFunction = true; else inMacro = true;
+        requireArgsSpecialVariable = false;
     nameExp = IdentifierOrStringLiteral();
 name = (nameExp instanceof StringLiteral)
                 ? ((StringLiteral) nameExp).getAsString()
@@ -2767,8 +2772,7 @@ isCatchAll = true;
       case EQUALS:{
         jj_consume_token(EQUALS);
         defValue = Expression();
-defNames.add(arg.image);
-                hasDefaults = true;
+hasDefaults = true;
         break;
         }
       default:
@@ -2784,7 +2788,7 @@ defNames.add(arg.image);
         jj_la1[65] = jj_gen;
         ;
       }
-if (catchAll != null) {
+if (catchAllParamName != null) {
                 {if (true) throw new ParseException(
                 "There may only be one \"catch-all\" parameter in a macro declaration, and it must be the last parameter.",
                 template, arg);}
@@ -2795,16 +2799,15 @@ if (catchAll != null) {
                     "\"Catch-all\" macro parameter may not have a default value.",
                     template, arg);}
                 }
-                catchAll = arg.image;
+                catchAllParamName = arg.image;
             } else {
-                argNames.add(arg.image);
                 if (hasDefaults && defValue == null) {
                     {if (true) throw new ParseException(
                                     "In a macro declaration, parameters without a default value "
                                     + "must all occur before the parameters with default values.",
                     template, arg);}
                 }
-                args.put(arg.image, defValue);
+                paramNamesWithDefault.put(arg.image, defValue);
             }
     }
     switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
@@ -2822,12 +2825,12 @@ if (catchAll != null) {
         iteratorBlockContexts = null;
         if (incompatibleImprovements >= _TemplateAPI.VERSION_INT_2_3_23) {
                 lastBreakableDirectiveNesting = breakableDirectiveNesting;
-                lastContiunableDirectiveNesting = continuableDirectiveNesting;
+                lastContinuableDirectiveNesting = continuableDirectiveNesting;
                 breakableDirectiveNesting = 0;
                 continuableDirectiveNesting = 0;
         } else {
             lastBreakableDirectiveNesting = 0; // Just to prevent uninitialized local variable error later
-            lastContiunableDirectiveNesting = 0;
+            lastContinuableDirectiveNesting = 0;
         }
     children = MixedContentElements();
     switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
@@ -2849,11 +2852,12 @@ if (!isFunction) {if (true) throw new ParseException("Expected macro end tag her
 iteratorBlockContexts = lastIteratorBlockContexts;
         if (incompatibleImprovements >= _TemplateAPI.VERSION_INT_2_3_23) {
             breakableDirectiveNesting = lastBreakableDirectiveNesting;
-            continuableDirectiveNesting = lastContiunableDirectiveNesting;
+            continuableDirectiveNesting = lastContinuableDirectiveNesting;
         }
 
         inMacro = inFunction = false;
-        Macro result = new Macro(name, argNames, args, catchAll, isFunction, children);
+        Macro result = new Macro(
+                name, paramNamesWithDefault, catchAllParamName, isFunction, requireArgsSpecialVariable, children);
         result.setLocation(template, start, end);
         template.addMacro(result);
         {if ("" != null) return result;}
@@ -2880,8 +2884,17 @@ CompressedBlock cb = new CompressedBlock(children);
     int pushedCtxCount = 0;
     start = jj_consume_token(UNIFIED_CALL);
     exp = Expression();
-if (exp instanceof Identifier || (exp instanceof Dot && ((Dot) exp).onlyHasIdentifiers())) {
-            startTagNameExp = exp;
+// To allow <@foo.bar?withArgs(...)>...</@foo.bar>, but we also remove superfluous (...):
+        Expression cleanedExp = exp;
+        if (cleanedExp instanceof MethodCall) {
+            Expression methodCallTarget = ((MethodCall) cleanedExp).getTarget();
+            if (methodCallTarget instanceof BuiltInsForCallables.with_argsBI) {
+                cleanedExp = ((BuiltInsForCallables.with_argsBI) methodCallTarget).target;
+            }
+        }
+
+        if (cleanedExp instanceof Identifier || (cleanedExp instanceof Dot && ((Dot) cleanedExp).onlyHasIdentifiers())) {
+            startTagNameExp = cleanedExp;
         } else {
             startTagNameExp = null;
         }
@@ -4804,12 +4817,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3_9()
- {
-    if (jj_scan_token(OPEN_PAREN)) return true;
-    return false;
-  }
-
   private boolean jj_3R_66()
  {
     if (jj_3R_81()) return true;
@@ -4836,6 +4843,13 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3_12()
+ {
+    if (jj_scan_token(ID)) return true;
+    if (jj_scan_token(EQUALS)) return true;
+    return false;
+  }
+
   private boolean jj_3R_64()
  {
     if (jj_3R_79()) return true;
@@ -4854,12 +4868,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_43()
- {
-    if (jj_scan_token(ID)) return true;
-    return false;
-  }
-
   private boolean jj_3R_60()
  {
     Token xsp;
@@ -4904,10 +4912,9 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3_12()
+  private boolean jj_3R_43()
  {
     if (jj_scan_token(ID)) return true;
-    if (jj_scan_token(EQUALS)) return true;
     return false;
   }
 
@@ -4951,34 +4958,32 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_75()
+  private boolean jj_3R_88()
  {
-    if (jj_3R_89()) return true;
+    if (jj_scan_token(OPEN_PAREN)) return true;
+    if (jj_3R_98()) return true;
+    if (jj_scan_token(CLOSE_PAREN)) return true;
     return false;
   }
 
-  private boolean jj_3R_79()
+  private boolean jj_3R_75()
  {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_scan_token(97)) {
-    jj_scanpos = xsp;
-    if (jj_scan_token(98)) return true;
-    }
+    if (jj_3R_89()) return true;
     return false;
   }
 
-  private boolean jj_3R_88()
+  private boolean jj_3R_74()
  {
-    if (jj_scan_token(OPEN_PAREN)) return true;
-    if (jj_3R_98()) return true;
-    if (jj_scan_token(CLOSE_PAREN)) return true;
+    if (jj_3R_88()) return true;
     return false;
   }
 
-  private boolean jj_3R_74()
+  private boolean jj_3R_114()
  {
-    if (jj_3R_88()) return true;
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_scan_token(130)) jj_scanpos = xsp;
+    if (jj_3R_113()) return true;
     return false;
   }
 
@@ -4994,6 +4999,17 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3R_79()
+ {
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_scan_token(97)) {
+    jj_scanpos = xsp;
+    if (jj_scan_token(98)) return true;
+    }
+    return false;
+  }
+
   private boolean jj_3R_61()
  {
     Token xsp;
@@ -5035,6 +5051,25 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3R_112()
+ {
+    if (jj_3R_113()) return true;
+    Token xsp;
+    while (true) {
+      xsp = jj_scanpos;
+      if (jj_3R_114()) { jj_scanpos = xsp; break; }
+    }
+    return false;
+  }
+
+  private boolean jj_3R_100()
+ {
+    if (jj_scan_token(OPEN_PAREN)) return true;
+    if (jj_3R_98()) return true;
+    if (jj_scan_token(CLOSE_PAREN)) return true;
+    return false;
+  }
+
   private boolean jj_3_2()
  {
     Token xsp;
@@ -5049,11 +5084,16 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_100()
+  private boolean jj_3_17()
  {
-    if (jj_scan_token(OPEN_PAREN)) return true;
-    if (jj_3R_98()) return true;
-    if (jj_scan_token(CLOSE_PAREN)) return true;
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_scan_token(79)) jj_scanpos = xsp;
+    xsp = jj_scanpos;
+    if (jj_scan_token(77)) {
+    jj_scanpos = xsp;
+    if (jj_scan_token(76)) return true;
+    }
     return false;
   }
 
@@ -5068,20 +5108,11 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_114()
+  private boolean jj_3R_109()
  {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(130)) jj_scanpos = xsp;
-    if (jj_3R_113()) return true;
-    return false;
-  }
-
-  private boolean jj_3R_83()
- {
-    if (jj_scan_token(OPEN_BRACKET)) return true;
-    if (jj_3R_98()) return true;
-    if (jj_scan_token(CLOSE_BRACKET)) return true;
+    if (jj_3R_112()) jj_scanpos = xsp;
     return false;
   }
 
@@ -5100,33 +5131,28 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_112()
+  private boolean jj_3R_83()
  {
-    if (jj_3R_113()) return true;
-    Token xsp;
-    while (true) {
-      xsp = jj_scanpos;
-      if (jj_3R_114()) { jj_scanpos = xsp; break; }
-    }
+    if (jj_scan_token(OPEN_BRACKET)) return true;
+    if (jj_3R_98()) return true;
+    if (jj_scan_token(CLOSE_BRACKET)) return true;
     return false;
   }
 
-  private boolean jj_3_17()
+  private boolean jj_3R_87()
  {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_scan_token(79)) jj_scanpos = xsp;
-    xsp = jj_scanpos;
-    if (jj_scan_token(77)) {
-    jj_scanpos = xsp;
-    if (jj_scan_token(76)) return true;
-    }
+    if (jj_scan_token(OPEN_BRACKET)) return true;
+    if (jj_3R_29()) return true;
+    if (jj_scan_token(CLOSE_BRACKET)) return true;
     return false;
   }
 
-  private boolean jj_3_7()
+  private boolean jj_3R_111()
  {
-    if (jj_scan_token(OR)) return true;
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_scan_token(130)) jj_scanpos = xsp;
+    if (jj_3R_29()) return true;
     return false;
   }
 
@@ -5141,25 +5167,34 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_109()
+  private boolean jj_3R_29()
  {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_112()) jj_scanpos = xsp;
+    if (jj_3R_33()) return true;
     return false;
   }
 
-  private boolean jj_3R_87()
+  private boolean jj_3_7()
+ {
+    if (jj_scan_token(OR)) return true;
+    return false;
+  }
+
+  private boolean jj_3R_108()
  {
-    if (jj_scan_token(OPEN_BRACKET)) return true;
     if (jj_3R_29()) return true;
-    if (jj_scan_token(CLOSE_BRACKET)) return true;
+    Token xsp;
+    while (true) {
+      xsp = jj_scanpos;
+      if (jj_3R_111()) { jj_scanpos = xsp; break; }
+    }
     return false;
   }
 
-  private boolean jj_3R_29()
+  private boolean jj_3R_98()
  {
-    if (jj_3R_33()) return true;
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_3R_108()) jj_scanpos = xsp;
     return false;
   }
 
@@ -5170,33 +5205,30 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_111()
+  private boolean jj_3_16()
  {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_scan_token(130)) jj_scanpos = xsp;
-    if (jj_3R_29()) return true;
+    if (jj_scan_token(HASH_INTERPOLATION_OPENING)) return true;
     return false;
   }
 
-  private boolean jj_3R_33()
+  private boolean jj_3_15()
  {
-    if (jj_3R_40()) return true;
     Token xsp;
-    while (true) {
-      xsp = jj_scanpos;
-      if (jj_3R_41()) { jj_scanpos = xsp; break; }
+    xsp = jj_scanpos;
+    if (jj_scan_token(82)) {
+    jj_scanpos = xsp;
+    if (jj_scan_token(84)) return true;
     }
     return false;
   }
 
-  private boolean jj_3R_108()
+  private boolean jj_3R_33()
  {
-    if (jj_3R_29()) return true;
+    if (jj_3R_40()) return true;
     Token xsp;
     while (true) {
       xsp = jj_scanpos;
-      if (jj_3R_111()) { jj_scanpos = xsp; break; }
+      if (jj_3R_41()) { jj_scanpos = xsp; break; }
     }
     return false;
   }
@@ -5207,26 +5239,12 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3_6()
- {
-    if (jj_scan_token(AND)) return true;
-    return false;
-  }
-
   private boolean jj_3R_38()
  {
     if (jj_scan_token(PLUS)) return true;
     return false;
   }
 
-  private boolean jj_3R_98()
- {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_108()) jj_scanpos = xsp;
-    return false;
-  }
-
   private boolean jj_3_1()
  {
     Token xsp;
@@ -5238,27 +5256,9 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3_16()
- {
-    if (jj_scan_token(HASH_INTERPOLATION_OPENING)) return true;
-    return false;
-  }
-
-  private boolean jj_3_15()
- {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_scan_token(82)) {
-    jj_scanpos = xsp;
-    if (jj_scan_token(84)) return true;
-    }
-    return false;
-  }
-
-  private boolean jj_3R_51()
+  private boolean jj_3_6()
  {
     if (jj_scan_token(AND)) return true;
-    if (jj_3R_50()) return true;
     return false;
   }
 
@@ -5274,17 +5274,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_40()
- {
-    if (jj_3R_50()) return true;
-    Token xsp;
-    while (true) {
-      xsp = jj_scanpos;
-      if (jj_3R_51()) { jj_scanpos = xsp; break; }
-    }
-    return false;
-  }
-
   private boolean jj_3R_99()
  {
     Token xsp;
@@ -5317,6 +5306,13 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3R_51()
+ {
+    if (jj_scan_token(AND)) return true;
+    if (jj_3R_50()) return true;
+    return false;
+  }
+
   private boolean jj_3R_28()
  {
     if (jj_3R_31()) return true;
@@ -5342,9 +5338,14 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3_5()
+  private boolean jj_3R_40()
  {
-    if (jj_3R_28()) return true;
+    if (jj_3R_50()) return true;
+    Token xsp;
+    while (true) {
+      xsp = jj_scanpos;
+      if (jj_3R_51()) { jj_scanpos = xsp; break; }
+    }
     return false;
   }
 
@@ -5366,21 +5367,34 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3_13()
+ {
+    if (jj_scan_token(OPEN_PAREN)) return true;
+    return false;
+  }
+
+  private boolean jj_3_5()
+ {
+    if (jj_3R_28()) return true;
+    return false;
+  }
+
   private boolean jj_3R_58()
  {
     if (jj_scan_token(MINUS)) return true;
     return false;
   }
 
-  private boolean jj_3R_106()
+  private boolean jj_3_14()
  {
-    if (jj_3R_28()) return true;
+    if (jj_scan_token(ID)) return true;
+    if (jj_scan_token(EQUALS)) return true;
     return false;
   }
 
-  private boolean jj_3_13()
+  private boolean jj_3R_106()
  {
-    if (jj_scan_token(OPEN_PAREN)) return true;
+    if (jj_3R_28()) return true;
     return false;
   }
 
@@ -5396,12 +5410,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_105()
- {
-    if (jj_scan_token(DOT_DOT_ASTERISK)) return true;
-    return false;
-  }
-
   private boolean jj_3R_91()
  {
     if (jj_scan_token(EXISTS)) return true;
@@ -5414,21 +5422,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_104()
- {
-    if (jj_scan_token(DOT_DOT_LESS)) return true;
-    return false;
-  }
-
-  private boolean jj_3R_93()
- {
-    if (jj_scan_token(DOT_DOT)) return true;
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_106()) jj_scanpos = xsp;
-    return false;
-  }
-
   private boolean jj_3R_52()
  {
     if (jj_scan_token(COMMA)) return true;
@@ -5436,13 +5429,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3_14()
- {
-    if (jj_scan_token(ID)) return true;
-    if (jj_scan_token(EQUALS)) return true;
-    return false;
-  }
-
   private boolean jj_3R_35()
  {
     if (jj_3R_43()) return true;
@@ -5463,27 +5449,30 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_110()
+  private boolean jj_3R_105()
  {
-    if (jj_3R_29()) return true;
+    if (jj_scan_token(DOT_DOT_ASTERISK)) return true;
     return false;
   }
 
-  private boolean jj_3R_92()
+  private boolean jj_3R_104()
  {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_104()) {
-    jj_scanpos = xsp;
-    if (jj_3R_105()) return true;
-    }
-    if (jj_3R_28()) return true;
+    if (jj_scan_token(DOT_DOT_LESS)) return true;
     return false;
   }
 
-  private boolean jj_3R_59()
+  private boolean jj_3R_110()
  {
-    if (jj_scan_token(EXCLAM)) return true;
+    if (jj_3R_29()) return true;
+    return false;
+  }
+
+  private boolean jj_3R_93()
+ {
+    if (jj_scan_token(DOT_DOT)) return true;
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_3R_106()) jj_scanpos = xsp;
     return false;
   }
 
@@ -5498,26 +5487,21 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_54()
+  private boolean jj_3R_59()
  {
-    Token xsp;
-    if (jj_3R_59()) return true;
-    while (true) {
-      xsp = jj_scanpos;
-      if (jj_3R_59()) { jj_scanpos = xsp; break; }
-    }
-    if (jj_3R_55()) return true;
+    if (jj_scan_token(EXCLAM)) return true;
     return false;
   }
 
-  private boolean jj_3R_78()
+  private boolean jj_3R_92()
  {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_3R_92()) {
+    if (jj_3R_104()) {
     jj_scanpos = xsp;
-    if (jj_3R_93()) return true;
+    if (jj_3R_105()) return true;
     }
+    if (jj_3R_28()) return true;
     return false;
   }
 
@@ -5547,12 +5531,15 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_62()
+  private boolean jj_3R_54()
  {
-    if (jj_3R_28()) return true;
     Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_78()) jj_scanpos = xsp;
+    if (jj_3R_59()) return true;
+    while (true) {
+      xsp = jj_scanpos;
+      if (jj_3R_59()) { jj_scanpos = xsp; break; }
+    }
+    if (jj_3R_55()) return true;
     return false;
   }
 
@@ -5566,6 +5553,17 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3R_78()
+ {
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_3R_92()) {
+    jj_scanpos = xsp;
+    if (jj_3R_93()) return true;
+    }
+    return false;
+  }
+
   private boolean jj_3R_80()
  {
     if (jj_scan_token(OPENING_CURLY_BRACKET)) return true;
@@ -5587,12 +5585,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_46()
- {
-    if (jj_3R_55()) return true;
-    return false;
-  }
-
   private boolean jj_3R_30()
  {
     Token xsp;
@@ -5604,15 +5596,24 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_45()
+  private boolean jj_3R_46()
  {
-    if (jj_3R_54()) return true;
+    if (jj_3R_55()) return true;
     return false;
   }
 
-  private boolean jj_3R_44()
+  private boolean jj_3R_62()
  {
-    if (jj_3R_53()) return true;
+    if (jj_3R_28()) return true;
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_3R_78()) jj_scanpos = xsp;
+    return false;
+  }
+
+  private boolean jj_3R_45()
+ {
+    if (jj_3R_54()) return true;
     return false;
   }
 
@@ -5652,17 +5653,9 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_36()
+  private boolean jj_3R_44()
  {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_44()) {
-    jj_scanpos = xsp;
-    if (jj_3R_45()) {
-    jj_scanpos = xsp;
-    if (jj_3R_46()) return true;
-    }
-    }
+    if (jj_3R_53()) return true;
     return false;
   }
 
@@ -5678,6 +5671,20 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3R_36()
+ {
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_3R_44()) {
+    jj_scanpos = xsp;
+    if (jj_3R_45()) {
+    jj_scanpos = xsp;
+    if (jj_3R_46()) return true;
+    }
+    }
+    return false;
+  }
+
   private boolean jj_3R_96()
  {
     if (jj_scan_token(FALSE)) return true;
@@ -5736,13 +5743,6 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
-  private boolean jj_3R_85()
- {
-    if (jj_scan_token(DOT)) return true;
-    if (jj_scan_token(ID)) return true;
-    return false;
-  }
-
   private boolean jj_3R_84()
  {
     if (jj_scan_token(OPEN_PAREN)) return true;
@@ -5786,6 +5786,13 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3R_85()
+ {
+    if (jj_scan_token(DOT)) return true;
+    if (jj_scan_token(ID)) return true;
+    return false;
+  }
+
   private boolean jj_3R_56()
  {
     if (jj_3R_62()) return true;
@@ -5813,6 +5820,12 @@ TemplateElement root = children.asSingleElement();
     return false;
   }
 
+  private boolean jj_3_9()
+ {
+    if (jj_scan_token(OPEN_PAREN)) return true;
+    return false;
+  }
+
   private boolean jj_3R_68()
  {
     if (jj_3R_83()) return true;
diff --git a/freemarker/core/IteratorBlock.java b/freemarker/core/IteratorBlock.java
index 25c27a6..c8de087 100644
--- a/freemarker/core/IteratorBlock.java
+++ b/freemarker/core/IteratorBlock.java
@@ -489,6 +489,7 @@ final class IteratorBlock extends TemplateElement {
             return visibleLoopVarName.equals(visibleLoopVar1Name) || visibleLoopVarName.equals(loopVar2Name);
         }
 
+        @Override
         public TemplateModel getLocalVariable(String name) {
             String visibleLoopVar1Name = this.visibleLoopVar1Name; // Not this.loopVar1Name!
             if (visibleLoopVar1Name == null) {
@@ -501,7 +502,7 @@ final class IteratorBlock extends TemplateElement {
                     case 0:
                         return loopVar1Value != null ? loopVar1Value
                                 : getTemplate().getConfiguration().getFallbackOnNullLoopVariable()
-                                        ? null : NullTemplateModel.INSTANCE;
+                                        ? null : TemplateNullModel.INSTANCE;
                     case 6: 
                         if (name.endsWith(LOOP_STATE_INDEX)) {
                             return new SimpleNumber(index);
@@ -518,12 +519,13 @@ final class IteratorBlock extends TemplateElement {
             if (name.equals(loopVar2Name)) {
                 return loopVar2Value != null ? loopVar2Value
                         : getTemplate().getConfiguration().getFallbackOnNullLoopVariable()
-                                ? null : NullTemplateModel.INSTANCE;
+                                ? null : TemplateNullModel.INSTANCE;
             }
             
             return null;
         }
         
+        @Override
         public Collection<String> getLocalVariableNames() {
             String visibleLoopVar1Name = this.visibleLoopVar1Name; // Not this.loopVar1Name!
             if (visibleLoopVar1Name != null) {
diff --git a/freemarker/ext/jsp/JspTagModelBase.java b/freemarker/ext/jsp/JspTagModelBase.java
index 26f8196..3d60cbe 100644
--- a/freemarker/ext/jsp/JspTagModelBase.java
+++ b/freemarker/ext/jsp/JspTagModelBase.java
@@ -118,7 +118,7 @@ class JspTagModelBase {
                                 " (declared type: ", new _DelayedShortClassName(setterType)
                                 + ", actual value's type: ",
                                 (argArray[0] != null
-                                        ? (Object) new _DelayedShortClassName(argArray[0].getClass()) : "Null"),
+                                        ? new _DelayedShortClassName(argArray[0].getClass()) : "Null"),
                                 "). See cause exception for the more specific cause...");
                         if (e instanceof IllegalArgumentException && !(setterType.isAssignableFrom(String.class))
                                 && argArray[0] != null && argArray[0] instanceof String) {
diff --git a/freemarker/ext/jsp/TagTransformModel.java b/freemarker/ext/jsp/TagTransformModel.java
index 625a547..78531cd 100644
--- a/freemarker/ext/jsp/TagTransformModel.java
+++ b/freemarker/ext/jsp/TagTransformModel.java
@@ -59,6 +59,7 @@ class TagTransformModel extends JspTagModelBase implements TemplateTransformMode
         isTryCatchFinally = TryCatchFinally.class.isAssignableFrom(tagClass);
     }
     
+    @Override
     public Writer getWriter(Writer out, Map args) throws TemplateModelException {
         try {
             Tag tag = (Tag) getTagInstance();
@@ -334,6 +335,7 @@ class TagTransformModel extends JspTagModelBase implements TemplateTransformMode
             return pageContext;
         }
         
+        @Override
         public int onStart()
         throws TemplateModelException {
             try {
@@ -372,6 +374,7 @@ class TagTransformModel extends JspTagModelBase implements TemplateTransformMode
             }
         }
         
+        @Override
         public int afterBody()
         throws TemplateModelException {
             try {
@@ -404,6 +407,7 @@ class TagTransformModel extends JspTagModelBase implements TemplateTransformMode
             }
         }
         
+        @Override
         public void onError(Throwable t) throws Throwable {
             if (isTryCatchFinally) {
                 ((TryCatchFinally) tag).doCatch(t);
diff --git a/freemarker/log/Logger.java b/freemarker/log/Logger.java
index 35e31cb..f931011 100644
--- a/freemarker/log/Logger.java
+++ b/freemarker/log/Logger.java
@@ -552,6 +552,7 @@ public abstract class Logger {
         try {
             return (String) AccessController.doPrivileged(
                     new PrivilegedAction() {
+                        @Override
                         public Object run() {
                             return System.getProperty(key, null);
                         }
diff --git a/freemarker/log/_Log4jLoggerFactory.java b/freemarker/log/_Log4jLoggerFactory.java
index a167f05..85f7a9a 100644
--- a/freemarker/log/_Log4jLoggerFactory.java
+++ b/freemarker/log/_Log4jLoggerFactory.java
@@ -27,89 +27,89 @@ import org.apache.logging.log4j.LogManager;
  * compliance issues. FreeMarker developers only: treat this class as package-visible.
  */
 public class _Log4jLoggerFactory implements LoggerFactory {
-	public Logger getLogger(String name) {
-		return new Log4jLogger(LogManager.getLogger(name));
-	}
+    @Override
+    public Logger getLogger(String name) {
+        return new Log4jLogger(LogManager.getLogger(name));
+    }
 
     private static class Log4jLogger
     extends
         Logger {
-		private final org.apache.logging.log4j.core.Logger logger;
-        
+        private final org.apache.logging.log4j.core.Logger logger;
+
         Log4jLogger(org.apache.logging.log4j.Logger logger) {
-			this.logger = (org.apache.logging.log4j.core.Logger)logger;
+            this.logger = (org.apache.logging.log4j.core.Logger)logger;
         }
         
         @Override
         public void debug(String message) {
-			logger.logIfEnabled(
-				_FQCN, Level.DEBUG, null, (Object)message, null);
+            logger.logIfEnabled(
+                _FQCN, Level.DEBUG, null, (Object)message, null);
         }
 
         @Override
         public void debug(String message, Throwable t) {
-			logger.logIfEnabled(_FQCN, Level.DEBUG, null, (Object)message, t);
+            logger.logIfEnabled(_FQCN, Level.DEBUG, null, (Object)message, t);
         }
 
         @Override
         public void error(String message) {
-			logger.logIfEnabled(
-				_FQCN, Level.ERROR, null, (Object)message, null);
+            logger.logIfEnabled(
+                _FQCN, Level.ERROR, null, (Object)message, null);
         }
 
         @Override
         public void error(String message, Throwable t) {
-			logger.logIfEnabled(_FQCN, Level.ERROR, null, (Object)message, t);
+            logger.logIfEnabled(_FQCN, Level.ERROR, null, (Object)message, t);
         }
 
         @Override
         public void info(String message) {
-			logger.logIfEnabled(_FQCN, Level.INFO, null, (Object)message, null);
+            logger.logIfEnabled(_FQCN, Level.INFO, null, (Object)message, null);
         }
 
         @Override
         public void info(String message, Throwable t) {
-			logger.logIfEnabled(_FQCN, Level.INFO, null, (Object)message, t);
+            logger.logIfEnabled(_FQCN, Level.INFO, null, (Object)message, t);
         }
 
         @Override
         public void warn(String message) {
-			logger.logIfEnabled(_FQCN, Level.WARN, null, (Object)message, null);
+            logger.logIfEnabled(_FQCN, Level.WARN, null, (Object)message, null);
         }
 
         @Override
         public void warn(String message, Throwable t) {
-			logger.logIfEnabled(_FQCN, Level.WARN, null, (Object)message, t);
+            logger.logIfEnabled(_FQCN, Level.WARN, null, (Object)message, t);
         }
 
         @Override
         public boolean isDebugEnabled() {
-			return logger.isDebugEnabled();
+            return logger.isDebugEnabled();
         }
 
         @Override
         public boolean isInfoEnabled() {
-			return logger.isInfoEnabled();
+            return logger.isInfoEnabled();
         }
 
         @Override
         public boolean isWarnEnabled() {
-			return logger.isWarnEnabled();
+            return logger.isWarnEnabled();
         }
 
         @Override
         public boolean isErrorEnabled() {
-			return logger.isErrorEnabled();
+            return logger.isErrorEnabled();
         }
 
         @Override
         public boolean isFatalEnabled() {
-			return logger.isFatalEnabled();
+            return logger.isFatalEnabled();
         }
 
-		private static final String _FQCN = Log4jLogger.class.getName();
-
-	}
+        private static final String _FQCN = Log4jLogger.class.getName();
 
+    }
 }
 /* @generated */
\ No newline at end of file
